### PR TITLE
Move handlers into their own file and clean up poutine namespace

### PR DESCRIPTION
**Pull request by eb8680**
Part of #826 

This PR creates a file `pyro/poutine/handlers.py` with the definitions that used to be in `pyro/poutine/__init__.py` to clean up the `pyro.poutine` namespace.  It also removes a couple of `Messenger`s in `pyro/primitives.py` that should have been handlers.

### CODE REVIEW

**Comment by eb8680** [comment]
> assuming this is purely refactoring

Yep, apart from the sphinx changes

**Comment by eb8680** [comment]
Should be good to review/merge, I'll change docstring contents in a followup PR.

File: docs/source/poutine.rst [hunk 1]
```diff
@@ -8,13 +8,28 @@ programs. Pyro's inference algorithms are all built by applying these handlers t
...
+
+.. automodule:: pyro.poutine.trace
+    :members:
+    :undoc-members:
+    :show-inheritance:
 
 Messengers
 -----------
 
 .. include:: pyro.poutine.txt
```
Comment (jpchen) [inline docs/source/poutine.rst]: not part of this PR, but this file should technically be changed to `pyro.poutine.messenger`

**Comment by fritzo** [comment]
This definitely improves our code organization, but I think we should update our sphinx docs. I'd like to be able to refer to e.g. `poutine.trace` as `pyro.poutine.trace`. To do this I believe we'll need to replace the `automodule:: pyro.poutine.trace` with
```
.. autoclass pyro.poutine.trace
    :members:
    :undoc-members:
    :show-inheritance:
```
and similar with a number of our other imported handlers and messengers like `pyro.poutine.do` (or if you prefer, simply `autoclass:: pyro.do`. This is fine to address in a follow-up PR.

**Comment by eb8680** [comment]
> To do this I believe we'll need to replace the `automodule:: pyro.poutine.trace`

I moved `trace.py` to `trace_struct.py`, does that help with this?

**Comment by fritzo** [comment]
> I moved trace.py to trace_struct.py

That helps :smile: but it is orthogonal to the issue I described with sphinx. The issue is that when we use `automodule:: pyro.foo` sphinx determines the fully qualified name of each thing `x` in `foo` by examining `x.__module__`. Since we've moved `trace` into `pyro.poutine.handlers`, sphinx determines its fully qualified name to be `pyro.poutine.handlers.trace`. While this is technically correct, we prefer to  call it `pyro.poutine.trace`. To resolve this issue we can instruct sphinx of our desired name by replacing `automodule` with one entry per thing in that module, `autofunction` or `autoclass` depending on each thing. Sphinx will use whatever name we provide, so e.g. we can achieve `pyro.poutine.trace` by a one-liner
```
.. autofunction pyro.poutine.trace
```
It's a little less automatic, but we should be able to replace
```diff
- .. automodule pyro.poutine
+ .. autofunction pyro.poutine.trace
+ .. autofunction pyro.poutine.replay
+ .. autofunction pyro.poutine.lift
+ .. autofunction pyro.poutine.block
+ .. autofunction pyro.poutine.escape
+ .. autofunction pyro.poutine.condition
+ .. autofunction pyro.poutine.infer_config
+ .. autofunction pyro.poutine.scale
+ .. autofunction pyro.poutine.indep
+ .. autofunction pyro.poutine.enum
+ .. autofunction pyro.poutine.do
+ .. autofunction pyro.poutine.queue
```
Let me know offline if this seems wrong or confusing.
